File: Array/luobiao/ReverseLinkedList.java
/**
反转一个单链表。

示例:

输入: 1->2->3->4->5->NULL
输出: 5->4->3->2->1->NULL
进阶:
你可以迭代或递归地反转链表。你能否用两种方法解决这道题？

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/reverse-linked-list
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。

主要思路：通过链表的属性对调前一个和后一个的赋值关系，最后判断链表的下一个不为空结束，此处的pre即是所要的最终链表

*/

class Solution {
    public ListNode reverseList(ListNode head) {
        if(head == null) {
            return head;
        }

        ListNode pre = null;
        ListNode current = head;
        ListNode next = current.next;
        while(current != null) {
            current.next = pre;
            pre = current;
            current = next;
            next = (next == null ? null : next.next);
        }
        
        return pre;
    }
}

File: Array/wuchao/linkedlist/7-2 反转链表 递归实现/ReverseLinkedList.c
//
// Created by 吴超 on 2020/7/2.
//

#include <stddef.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode *reverse(struct ListNode *prev, struct ListNode *current);

struct ListNode *reverseList(struct ListNode *head) {
    return reverse(NULL, head);
}

struct ListNode *reverse(struct ListNode *prev, struct ListNode *current) {
    if (current == NULL) {
        return prev;
    }
    struct ListNode *next = current->next;
    current->next = prev;
    return reverse(current, next);
}
